src/api/procedures/setTransferRestrictions.ts

```diff
@@ -100,41 +137,77 @@ export async function prepareSetTransferRestrictions(
     });
   }
 
-  const maxTransferManagers = u32ToBigNumber(consts.statistics.maxTransferManagersPerAsset);
+  const maxTransferConditions = u32ToBigNumber(consts.statistics.maxTransferConditionsPerAsset);
   const finalCount = occupiedSlots.plus(newRestrictionAmount);
-  if (finalCount.gte(maxTransferManagers)) {
+  if (finalCount.gte(maxTransferConditions)) {
     throw new PolymeshError({
       code: ErrorCode.LimitExceeded,
       message: 'Cannot set more Transfer Restrictions than there are slots available',
       data: {
-        availableSlots: maxTransferManagers.minus(occupiedSlots),
+        availableSlots: maxTransferConditions.minus(occupiedSlots),
       },
     });
   }
 
-  const transactions = assembleBatchTransactions(
-    tuple(
-      {
-        transaction: statistics.removeTransferManager,
-        argsArray: restrictionsToRemove,
-      },
-      {
-        transaction: statistics.addTransferManager,
-        argsArray: restrictionsToAdd,
-      },
-      {
-        transaction: statistics.removeExemptedEntities,
-        argsArray: exemptionsToRemove,
-      },
-      {
-        transaction: statistics.addExemptedEntities,
-        argsArray: exemptionsToAdd,
-      }
-    )
+  const transactions = [];
+
+  const op =
+    type === TransferRestrictionType.Count
+      ? primitiveOpType(StatisticsOpType.Count, context)
+      : primitiveOpType(StatisticsOpType.Balance, context);
+
+  if (needStat) {
+    const newStat = primitiveStatisticsStatType(op, context);
+    currentStats.push(newStat);
+    currentStats.sort().reverse();
+
+    transactions.push(
+      checkTxType({
+        transaction: statistics.setActiveAssetStats,
+        args: [{ Ticker: rawTicker }, currentStats],
+      })
+    );
+
+    // Count stats need to be initialized manually
+    if (type === TransferRestrictionType.Count) {
+      const holders = await asset.balanceOf.entries(rawTicker);
+      const holderCount = new BigNumber(holders.length);
+      const secondKey = primitive2ndKey(context);
+      const stat = [
+        statUpdate(secondKey, bigNumberToU128(holderCount, context), context),
+      ] as BTreeSetStatUpdate;
+
+      transactions.push(
+        checkTxType({
+          transaction: statistics.batchUpdateAssetStats,
+          args: [{ Ticker: rawTicker }, newStat, stat],
+        })
+      );
+    }
+  }
+
+  transactions.push(
+    checkTxType({
+      transaction: statistics.setAssetTransferCompliance,
+      args: [{ Ticker: rawTicker }, conditions],
+    })
   );
 
-  this.addBatchTransaction({ transactions });
+  if (exemptions.length) {
+    const exemptedIds = await getExemptedIds(exemptions, context, ticker);
+    const exemptedScopeIds = exemptedIds.map(entityId => stringToIdentityId(entityId, context));
+    const btreeIds = scopeIdsToBtreeSetIdentityId(exemptedScopeIds, context);
```

Comment:
I sort inside here, so might be better to have other BTree methods use a similar pattern.